src/plugins/data/public/search/session/session_service.ts

```diff
@@ -275,6 +279,14 @@ export class SessionService {
     return !!sessionId && this.getSessionId() === sessionId;
   }
 
+  /**
+   * Should cache on client
+   * @param sessionId
+   */
+  public shouldCacheOnClient(sessionId?: string): boolean {
+    return !!sessionId && (this.state.get().cacheOnClient ?? false);
```

Comment:
Everything inside `this.state.get()` in any given moment is a state of a particular single session:

so I'd check if received `sessionId` is actually current `sessionId` otherwise `this.state.get().cacheOnClient` theoretically could belong to a different session